File: src/token.ml
open Base

type token =
  | INT of int
  | FLOAT of float
  | CHAR of string
  | STRING of string
  | GETTER of string
  | QUALIFIER of string
  | LOWER_NAME of string
  | UPPER_NAME of string
  | HOLE of int
  | PLUS
  | MINUS
  | TIMES
  | DIV
  | PERCENT
  | POWER
  | SHL
  | SHR
  | CARET
  | ANDAND
  | AND
  | OROR
  | LTE
  | LT
  | EQ
  | NEQ
  | GT
  | GTE
  | RANGE_I
  | RANGE_E
  | TILDE
  | IF
  | THEN
  | ELSE
  | TRUE
  | FALSE
  | TYPE
  | ALIAS
  | MODULE
  | PRIVATE
  | OPAQUE
  | EXTEND
  | BACKSLASH
  | ARROW
  | UNDERSCORE
  | LHOLE
  | LPAREN
  | RPAREN
  | LBRACE
  | RBRACE
  | LBRACKET
  | RBRACKET
  | COMMA
  | COLON
  | BANG
  | EQUALS
  | PIPE
  | SHEBANG
  | EOL
  | EOF
  [@@deriving show, compare]

File: src/token.ts
import {Loc} from './loc.ts';

export type Token = {type:TokenType, loc: Loc}

export type TokenType =
    | {type:'INT',int:number}
    | {type:'FLOAT',float:number}
    | {type:'CHAR',char:string}
    | {type:'STRING',string:string}
    | {type:'GETTER',field:string}
    | {type:'QUALIFIER',name:string}
    | {type:'LOWER_NAME',name:string}
    | {type:'UPPER_NAME',name:string}
    | {type:'HOLE',n:number}
    | {type:SimpleTokenType}

export type TokenTag = TokenType['type'];

export type SimpleTokenType =
    | 'PLUS'
    | 'MINUS'
    | 'TIMES'
    | 'DIV'
    | 'PERCENT'
    | 'POWER'
    | 'PLUSPLUS'
    | 'SHL'
    | 'SHR'
    | 'SHRU'
    | 'CARET'
    | 'ANDAND'
    | 'AND'
    | 'OROR'
    | 'LTE'
    | 'LT'
    | 'EQEQ'
    | 'NEQ'
    | 'GT'
    | 'GTE'
    | 'DOTDOT'
    | 'DOTDOTDOT'
    | 'TILDE'
    | 'CASE'
    | 'OF'
    | 'IF'
    | 'THEN'
    | 'ELSE'
    | 'TRUE'
    | 'FALSE'
    | 'TYPE'
    | 'ALIAS'
    | 'MODULE'
    | 'PRIVATE'
    | 'OPAQUE'
    | 'EXTEND'
    | 'BACKSLASH'
    | 'ARROW'
    | 'UNDERSCORE'
    | 'LHOLE'
    | 'LPAREN'
    | 'RPAREN'
    | 'LBRACE'
    | 'RBRACE'
    | 'LBRACKET'
    | 'RBRACKET'
    | 'PIPELINE'
    | 'COMMA'
    | 'COLON'
    | 'BANG'
    | 'EQ'
    | 'PIPE'
    | 'EOL'
    | 'EOF'
